src/streaming.rs

```diff
@@ -0,0 +1,270 @@
+use ::open_vcdiff_sys;
+
+use std::error;
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem;
+use std::os::raw::c_void;
+use std::slice;
+use std::result;
+
+#[derive(Debug,Copy,Clone,PartialEq,Eq)]
+pub struct Error(&'static str);
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        self.0
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "vcdiff error: {}", self.0)
+    }
+}
+
+pub type Result<T> = result::Result<T,Error>;
+
+pub trait Output {
+    fn append(&mut self, buffer: &[u8]);
+    fn clear(&mut self);
+    fn reserve(&mut self, bytes: usize);
+    fn size(&self) -> usize;
+}
+
+impl Output for Vec<u8> {
+    fn append(&mut self, buffer: &[u8]) {
+        self.extend_from_slice(buffer);
+    }
+
+    fn clear(&mut self) {
+        self.clear();
+    }
+
+    fn reserve(&mut self, bytes: usize) {
+        self.reserve(bytes);
+    }
+
+    fn size(&self) -> usize {
+        self.len()
+    }
+}
+
+// Set up some Rust functions to call from C which delegate to an Output
+
+unsafe extern "C" fn output_trait_append_cb(callback_pointer: *mut c_void, buffer: *const u8, buffer_size: usize) {
+    let mut callback_pointer: Box<&mut Output> = mem::transmute(callback_pointer);
+    let buffer: &[u8] = slice::from_raw_parts(buffer, buffer_size);
+    callback_pointer.append(buffer);
+}
+
+unsafe extern "C" fn output_trait_clear_cb(callback_pointer: *mut c_void) {
+    let mut callback_pointer: Box<&mut Output> = mem::transmute(callback_pointer);
+    callback_pointer.clear();
+}
+
+unsafe extern "C" fn output_trait_reserve_cb(callback_pointer: *mut c_void, bytes: usize) {
+    let mut callback_pointer: Box<&mut Output> = mem::transmute(callback_pointer);
+    callback_pointer.reserve(bytes);
+}
+
+unsafe extern "C" fn output_trait_size_cb(callback_pointer: *const c_void) -> usize {
+    let callback_pointer: Box<&Output> = mem::transmute(callback_pointer);
+    callback_pointer.size()
+}
+
+pub struct Decoder<'d> {
+    decoder: *mut c_void,
+
+    // decoder holds a borrow across the FFI boundary; model this as PhantomData
+    dictionary: PhantomData<&'d [u8]>,
+}
+
+impl<'d> Decoder<'d> {
+    pub fn new(dictionary: &'d [u8]) -> Decoder<'d> {
+        // make a new decoder
+        // this is C++ "operator new" under the hood
+        let decoder = unsafe { open_vcdiff_sys::new_decoder() };
+
+        // point it to the dictionary buffer
+        // this is a borrow we'll keep for the lifetime of the Decoder, so there's no need to copy it
+        unsafe { open_vcdiff_sys::decoder_start_decoding(decoder, dictionary.as_ptr() as *const i8, dictionary.len()) }
+
+        // Decoder is ready to use
+        Decoder{
+            decoder: decoder,
+            dictionary: PhantomData,
+        }
+    }
+
+    pub fn with_options(
```

Comment:
There's no easy way to add options with this interface, we'd need a builder or something.  But that's also a separate PR perhaps.